File: C#/Leetcode/String/RomanToInteger.cs
using System;
using System.Collections.Generic;

namespace LeetcodeSolutions.String
{
    // Leetcode 13 - https://leetcode.com/problems/roman-to-integer/
    // Submission Detail - https://leetcode.com/submissions/detail/123691683/
    // 2 pointers

    public class RomanToInteger
    {
        //static void Main(string[] args)
        //{
        //    Console.WriteLine($"Input: {"MMCDXLII"}, Output: {RomanToInt("MMCDXLII")}\tExpected:2442");
        //    Console.WriteLine($"Input: {"MMDCCLXXIII"}, Output: {RomanToInt("MMDCCLXXIII")}\tExpected:2773");
        //    Console.WriteLine($"Input: {"DCCCLXXVI"}, Output: {RomanToInt("DCCCLXXVI")}\tExpected:876");
        //    Console.WriteLine($"Input: {"CMLVII"}, Output: {RomanToInt("CMLVII")}\tExpected:957");
        //    Console.WriteLine($"Input: {"MMCCCXXVIII"},\tOutput: {RomanToInt("MMCCCXXVIII")},\tExpected:2328");
        //    Console.WriteLine($"Input: {"MMDCXCIX"}, Output: {RomanToInt("MMDCXCIX")}\tExpected:2699");

        //    Console.ReadLine();
        //}

        // 1-5000 Numerals - http://romannumerals.babuo.com/roman-numerals-1-5000
        private static Dictionary<char, int> values = new Dictionary<char, int>
            {
                { 'I', 1 },
                { 'V', 5 },
                { 'X', 10 },
                { 'L', 50 },
                { 'C', 100 },
                { 'D', 500 },
                { 'M', 1000 }
            };

        // Runtime : 156 ms
        // Tx = O(n) {n: n is the length of the input string}
        // Sx = O(1)
        // TODO: Add validation for allowed roman numerals combination
        public static int RomanToInt(string s)
        {
            int computedValue = 0;

            for (int i = 0; i < s.Length; i++)
            {
                if (!values.TryGetValue(s[i], out int val))
                    throw new ArgumentNullException("Invalid character found.");

                if (i == s.Length - 1 || values[s[i + 1]] <= val)
                {
                    computedValue += val;
                }
                else if (values[s[i + 1]] > val)
                {
                    computedValue += values[s[i + 1]] - val;
                    i++;
                }
            }

            return computedValue;
        }
    }
}

File: Java/Strings/RomanToInteger.java
package Strings;

import java.util.Map;

public class RomanToInteger {
    // Leetcode 13: https://leetcode.com/problems/roman-to-integer/description/
    // Submission Detail: https://leetcode.com/problems/roman-to-integer/submissions/1708930622

    //Example 1:
    //Input: s = "III"
    //Output: 3
    //Explanation: III = 3.

    //Example 2:
    //Input: s = "LVIII"
    //Output: 58
    //Explanation: L = 50, V= 5, III = 3.

    //Example 3:
    //Input: s = "MCMXCIV"
    //Output: 1994
    //Explanation: M = 1000, CM = 900, XC = 90 and IV = 4.

    private static final Map<Character, Integer> romanToIntMapping = Map.of(
        'I', 1,
        'V', 5,
        'X', 10,
        'L', 50,
        'C', 100,
        'D', 500,
        'M', 1000
    );

    public int romanToInt(String s) {
        int result=0;
        for(int index=0; index<s.length(); index++) {
            int currentCharValue = romanToIntMapping.get(s.charAt(index));
            int nextCharValue = index+1 < s.length() ? romanToIntMapping.get(s.charAt(index+1)) : 0;

            if(currentCharValue < nextCharValue) {
                result -= currentCharValue;
            } else {
                result += currentCharValue;
            }
        }

        return result;
    }
}
